capstone/src/main/java/com/google/step/search/ElasticsearchClient.java

```diff
@@ -55,21 +64,88 @@ public ElasticsearchClient(String hostname, short port) {
   }
 
   /**
-   * Given a {@code RestaurantHeader}, sends HTTP request to Elasticsearch server to add a document
-   * to the "restaurants" index representing a {@code restaurantHeader}.
-   * @param restaurantHeader a {@code RestaurantHeader}
+   * Given a {@link RestaurantHeader}, sends HTTP request to the search server to add a document
+   * that represents a {@link RestaurantHeader} to the "restaurants" index.
+   * @param restaurantHeader a {@link RestaurantHeader} that will be updated in the search index
    * @throws IOException when buildPutRequest() or execute() fails
    */
   public void updateRestaurantHeader(RestaurantHeader restaurantHeader) throws IOException {
     String restaurantKey = String.valueOf(restaurantHeader.getRestaurantKey());
+    List<String> requestPath = Arrays.asList("", RESTAURANTS, "_doc", restaurantKey);
 
+    String requestBody = gson.toJson(restaurantHeader);
+
+    HttpRequest request = buildElasticsearchHttpRequest("PUT", requestPath, requestBody);
+    request.execute();
+  }
+
+  /**
+   * Queries search server for {@link RestaurantHeader} fields that match {@code query} by either
+   * "name" or "cuisine" field.
+   * @param query valid query string
+   * @return list of {@link RestaurantHeader} objects sorted by descending relevance score
+   * @throws IOException if request cannot be made or executed properly
+   */
+  public List<RestaurantHeader> queryRestaurantHeaders(String query) throws IOException {
+    List<String> requestPath = Arrays.asList("", RESTAURANTS, "_search");
+
+    String requestBody = new JSONObject()
+        .put("query", new JSONObject()
+            .put("multi_match", new JSONObject()
+                .put("query", query)
+                .put("fields", new JSONArray(Arrays.asList("name", "cuisine")))))
+        .toString();
+
+    HttpRequest request = buildElasticsearchHttpRequest("POST", requestPath, requestBody);
+    HttpResponse response = request.execute();
+
+    return convertElasticsearchResponseBodyToHeaders(response);
+  }
+
+  /**
+   * Queries search server, matching everything possible.
+   * @return list of {@link RestaurantHeader} objects sorted arbitrarily
+   * @throws IOException if request cannot be made or executed properly
+   */
+  public List<RestaurantHeader> getRandomRestaurants() throws IOException {
+    List<String> requestPath = Arrays.asList("", "_search");
+
+    String requestBody = new JSONObject()
+        .put("query", new JSONObject()
+            .put("match_all", new JSONObject()))
+        .toString();
+
+    HttpRequest request = buildElasticsearchHttpRequest("POST", requestPath, requestBody);
+    HttpResponse response = request.execute();
+
+    return convertElasticsearchResponseBodyToHeaders(response);
+  }
+
+  private HttpRequest buildElasticsearchHttpRequest(String requestMethod, List<String> urlPath, String requestBody) throws IOException {
```

Comment:
Nit: add method documentation

Reply:
I typically leave documentation off of private methods, as they aren't user-facing.

Reply:
I see the point, but I agree with Katie it would be better to put it so we as the collaborators can get a better understanding and for yourself or others looking at the code in the future.

Reply:
Okay, I will add javadocs to the two private methods.

Reply:
Although, I will say that in most cases, code should be self-documenting. I typically try to reserve javadoc comments for a user who does not necessarily have access to the code, only the interface. If it is the case that my code is hard to read or understand, please let me know so that I can work on improving the readability. Javadocs should never be a replacement for poorly written code.

Reply:
I think your code is great and very clear. I hadn't heard of leaving javadocs off of private methods but it makes a lot of sense! Thanks for the tip.

Reply:
+1 to Josh's comments. Typically style guides don't require comments on private methods, although it's a good idea to add comments if the helper method is particularly complex/hard to understand.